dali/test/python/operator_2/test_warp_perspective.py

```diff
@@ -0,0 +1,372 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import nvidia.dali as dali
+import nvidia.dali.fn as fn
+import cv2
+import numpy as np
+import multiprocessing as mp
+import test_utils
+from nose2.tools import params
+
+NUM_THREADS = mp.cpu_count()
+DEV_ID = 0
+SEED = 1313
+
+
+def ocv_border_mode(border_mode):
+    if border_mode == "constant":
+        return cv2.BORDER_CONSTANT
+    elif border_mode == "replicate":
+        return cv2.BORDER_REPLICATE
+    elif border_mode == "reflect":
+        return cv2.BORDER_REFLECT
+    elif border_mode == "reflect_101":
+        return cv2.BORDER_REFLECT_101
+    elif border_mode == "wrap":
+        return cv2.BORDER_WRAP
+    else:
+        raise ValueError("Invalid border mode")
+
+
+def ocv_interp_type(interp_type):
+    if interp_type == "nearest":
+        return cv2.INTER_NEAREST
+    elif interp_type == "linear":
+        return cv2.INTER_LINEAR
+    elif interp_type == "cubic":
+        return cv2.INTER_CUBIC
+    else:
+        raise ValueError("Invalid interpolation type")
+
+
+def cv2_warp_perspective(
+    dst, img, matrix, layout, border_mode, interp_type, inverse_map, fill_value=None
+):
+    border_mode = ocv_border_mode(border_mode)
+    interp_type = ocv_interp_type(interp_type)
+    flags = interp_type
+
+    if fill_value is None:
+        fill_value = 0
+
+    if inverse_map:
+        flags = flags | cv2.WARP_INVERSE_MAP
+    if layout[-1] == "C":
+        dsize = (dst.shape[1], dst.shape[0])
+        if not isinstance(fill_value, tuple):
+            fill_value = tuple([fill_value for c in range(dst.shape[2])])
+        dst[:, :, :] = cv2.warpPerspective(
+            img, M=matrix, dsize=dsize, flags=flags, borderMode=border_mode, borderValue=fill_value
+        ).reshape(dst.shape)
+    else:
+        dsize = (dst.shape[2], dst.shape[1])
+        for c in range(img.shape[0]):
+            dst[c, :, :] = cv2.warpPerspective(
+                img[c, :, :],
+                M=matrix,
+                dsize=dsize,
+                flags=flags,
+                borderMode=border_mode,
+                borderValue=fill_value,
+            ).reshape(dst.shape[1:])
+
+
+def ref_func(img, matrix, size, layout, border_mode, interp_type, inverse_map, fill_value):
+    out_shape = list(img.shape)
+    if size is not None:
+        out_shape[layout.find("H")] = size[0]
+        out_shape[layout.find("W")] = size[1]
+    dst = np.zeros(out_shape, dtype=img.dtype)
+    if layout[0] == "F":
+        for f in range(0, img.shape[0]):
+            cv2_warp_perspective(
+                dst[f, :, :, :],
+                img[f, :, :, :],
+                matrix,
+                layout,
+                border_mode,
+                interp_type,
+                inverse_map,
+                fill_value,
+            )
```

Comment:
I have objections against using OpenCV as a reference. It has a different notion of pixel centers than we do. If we match OpenCV, it's actually a bad thing. Instead, when the perspective matrix is actually an affine transform matrix, WarpPerspective should match WarpAffine.